src/nxt_http_compression.c

```diff
@@ -0,0 +1,568 @@
+        .cops       = &nxt_comp_brotli_ops,
+    },
+#endif
+};
+
+static void print_compressor(const nxt_http_comp_compressor_t *c)
+{
+    printf("token    : %s\n", c->type->token.start);
+    printf("scheme   : %d\n", c->type->scheme);
+    printf("level    : %d\n", c->opts.level);
+    printf("min_len  : %ld\n", c->opts.min_len);
+}
+
+static void print_comp_config(size_t n)
+{
+    for (size_t i = 0; i < n; i++) {
+        nxt_http_comp_compressor_t *compr = enabled_compressors + i;
+
+        print_compressor(compr);
+        printf("\n");
+    }
+}
+
+bool
+nxt_http_comp_wants_compression(void)
+{
+    printf("%s: compression [%s]\n", __func__,
+           compressor_ctx.idx > 0 ? "true" : "false");
+    return compressor_ctx.idx;
+}
+
+size_t
+nxt_http_comp_bound(size_t size)
+{
+    nxt_http_comp_compressor_t        *compressor;
+    const nxt_http_comp_operations_t  *cops;
+
+    compressor = &enabled_compressors[compressor_ctx.idx];
+    cops = compressor->type->cops;
+
+    return cops->bound(&compressor_ctx.ctx, size);
+}
+
+ssize_t
+nxt_http_comp_compress(uint8_t *dst, size_t dst_size, const uint8_t *src,
+                       size_t src_size, bool last)
+{
+    nxt_http_comp_compressor_t        *compressor;
+    const nxt_http_comp_operations_t  *cops;
+
+    compressor = &enabled_compressors[compressor_ctx.idx];
+    cops = compressor->type->cops;
+
+    return cops->deflate(&compressor_ctx.ctx, src, src_size, dst, dst_size,
+                         last);
+}
+
+static nxt_uint_t
+nxt_http_comp_compressor_lookup_enabled(const nxt_str_t *token)
+{
+    if (token->start[0] == '*') {
+        return NXT_HTTP_COMP_SCHEME_IDENTITY;
+    }
+
+    for (size_t i = 0, n = nr_enabled_compressors; i < n; i++) {
+        if (nxt_strstr_eq(token, &enabled_compressors[i].type->token)) {
+            return i;
+        }
+    }
+
+    return NXT_HTTP_COMP_SCHEME_UNKNOWN;
+}
+
+/*
+ * We need to parse the 'Accept-Encoding` header as described by
+ * <https://www.rfc-editor.org/rfc/rfc9110.html#field.accept-encoding>
+ * which can take forms such as
+ *
+ *  Accept-Encoding: compress, gzip
+ *  Accept-Encoding:
+ *  Accept-Encoding: *
+ *  Accept-Encoding: compress;q=0.5, gzip;q=1.0
+ *  Accept-Encoding: gzip;q=1.0, identity;q=0.5, *;q=0
+ *
+ *  '*:q=0' means if the content being served has no 'Content-Coding'
+ *  matching an 'Accept-Encoding' entry then don't send any response.
+ *
+ * 'indentity;q=0' seems to basically mean the same thing...
+ */
+static nxt_int_t
+nxt_http_comp_select_compressor(const nxt_str_t *token)
+{
+    bool identity_allowed = true;
+    char *str;
+    char *saveptr;
+    double weight = 0.0;
+    nxt_int_t idx = 0;
+
+    str = strndup((char *)token->start, token->length);
+
+    for ( ; ; str = NULL) {
+        char *tkn, *qptr;
+        double qval = -1.0;
+        nxt_uint_t ecidx;
+        nxt_str_t enc;
+        nxt_http_comp_scheme_t scheme;
+
+        tkn = strtok_r(str, ", ", &saveptr);
```

Comment:
Hi @ac000,

strtok_r(3) isn't very good for this, since it will also split on ` `.

Let's take the following example, taken literally from RFC 9110:

```
Accept-Encoding: gzip;q=1.0, identity; q=0.5, *;q=0
```

One of the tokens would be `q=0.5`.

What you want is to delimit on `,`, and _after_ that trim whitespace.

Another problem of strtok_r(3) is that it merges adjacent delimiters, which would be bogus input and should be rejected.

strsep(3) doesn't have that issue.

You probably want:

```c
#define stpspn(s, reject)  (s + strspn(s, reject))

tkn = stpspn(strsep(&str, ","), " ");
weigth = stpsep(str, "; ");
```
With stpsep() being a function of mine similar to strsep(3).

Bad part: strsep(3) is a GNU extension.